File: MonoGameVBScript/GameHost.vb
' ===== NOTE ON GLOBAL IMPORTS =====
' Global imports for this hosting layer are defined in the VB.NET project file.
' See the `MonoGameVBScript.vbproj` file for details.

Public NotInheritable Class GameHost
    Inherits Game

    Private ReadOnly _graphics As GraphicsDeviceManager
    Private _spriteBatch As SpriteBatch
    Private _vbsEngine As VBScriptEngine

#Region "Public API Methods"
    Public Sub SetWindowSize(width As Integer, height As Integer)
        _graphics.PreferredBackBufferWidth = width
        _graphics.PreferredBackBufferHeight = height
        _graphics.ApplyChanges()
    End Sub

    Public ReadOnly Property ViewportWidth As Integer
        Get
            Return _graphics.PreferredBackBufferWidth
        End Get
    End Property

    Public ReadOnly Property ViewportHeight As Integer
        Get
            Return _graphics.PreferredBackBufferHeight
        End Get
    End Property

    Public ReadOnly Property SpriteBatch As SpriteBatch
        Get
            Return _spriteBatch
        End Get
    End Property
#End Region

    Private Sub New()
        _graphics = New GraphicsDeviceManager(Me)
        Content.RootDirectory = "Content"
        IsMouseVisible = True
    End Sub

    Protected Overrides Sub Initialize()
        ' Initialize VBScript engine, and expose API in LoadContent()
        _vbsEngine = New VBScriptEngine
        MediaPlayer.IsRepeating = True

        MyBase.Initialize()
    End Sub

    Protected Overrides Sub LoadContent()
        ArgumentNullException.ThrowIfNull(_vbsEngine)
        _spriteBatch = New SpriteBatch(GraphicsDevice)
        _vbsEngine.AddHostObject("game", Me)
        _vbsEngine.LoadRequiredContent(Content)
        _vbsEngine.Execute(IO.File.ReadAllText("GameCore.vbs"))
        Try
            _vbsEngine.Invoke("Initialize")
        Catch ex As Exception
            MsgBox($"Error in script Initialize: {ex.Message}")
            [Exit]()
        End Try

        MyBase.LoadContent()
    End Sub

    Protected Overrides Sub Update(gameTime As GameTime)
        If GamePad.GetState(PlayerIndex.One).Buttons.Back = ButtonState.Pressed OrElse
           Keyboard.GetState().IsKeyDown(Keys.Escape) Then [Exit]()

        ' Call script Update function
        Try
            _vbsEngine?.Invoke("Update", gameTime.ElapsedGameTime.TotalSeconds)
        Catch ex As Exception
            MsgBox($"Error in script Update: {ex.Message}")
            [Exit]()
        End Try

        MyBase.Update(gameTime)
    End Sub

    Protected Overrides Sub Draw(gameTime As GameTime)
        ' Call script Draw function
        Try
            _vbsEngine?.Invoke("Draw", gameTime.ElapsedGameTime.TotalSeconds)
        Catch ex As Exception
            MsgBox($"Error in script Draw: {ex.Message}")
            [Exit]()
        End Try

        MyBase.Draw(gameTime)
    End Sub

    Protected Overrides Sub Dispose(disposing As Boolean)
        If disposing Then
            _spriteBatch?.Dispose()
            _vbsEngine?.Dispose()
        End If

        MyBase.Dispose(disposing)
    End Sub

    Friend Shared Sub Main()
        Using host As New GameHost
            host.Run()
        End Using
    End Sub
End Class

File: MonoGameJScript/GameHost.cs
using MessageBox = System.Windows.Forms.MessageBox;

namespace MonoGameJScript;

/// <summary>
/// Game host that exposes API to JScript engine.
/// </summary>
public sealed class GameHost : Game
{
    private readonly GraphicsDeviceManager _graphics;
    private SpriteBatch? _spriteBatch;
    private JScriptEngine? _jsEngine;

    public GameHost()
    {
        _graphics = new GraphicsDeviceManager(this);
        Content.RootDirectory = "Content";
        IsMouseVisible = true;

    }

    #region Public API Methods
    public void SetWindowSize(int width, int height)
    {
        _graphics.PreferredBackBufferWidth = width;
        _graphics.PreferredBackBufferHeight = height;
        _graphics.ApplyChanges();
    }
    public int ViewportWidth => GraphicsDevice.Viewport.Width;
    public int ViewportHeight => GraphicsDevice.Viewport.Height;
    public SpriteBatch SpriteBatch => _spriteBatch!;
    #endregion

    protected override void Initialize()
    {
        // Initialize JScript engine, and expose API in LoadContent()
        _jsEngine = new JScriptEngine();
        MediaPlayer.IsRepeating = true;

        base.Initialize();
    }

    protected override void LoadContent()
    {
        ArgumentNullException.ThrowIfNull(_jsEngine);
        _spriteBatch = new SpriteBatch(GraphicsDevice);
        _jsEngine.AddHostObject("game", this);
        _jsEngine.LoadRequiredContent(Content);

        _jsEngine.Execute(File.ReadAllText("GameCore.js"));
        try
        {
            _jsEngine.Invoke("Initialize");
        }
        catch (Exception ex)
        {
            MessageBox.Show($"Error in script Initialize: {ex.Message}");
            Exit();
        }

        base.LoadContent();
    }

    protected override void Update(GameTime gameTime)
    {
        if (GamePad.GetState(PlayerIndex.One).Buttons.Back == ButtonState.Pressed ||
            Keyboard.GetState().IsKeyDown(Keys.Escape)) Exit();

        // Call script Update function
        try
        {
            _jsEngine?.Invoke("Update", gameTime.ElapsedGameTime.TotalSeconds);
        }
        catch (Exception ex)
        {
            MessageBox.Show($"Error in script Update: {ex.Message}");
            Exit();
        }

        base.Update(gameTime);
    }

    protected override void Draw(GameTime gameTime)
    {
        // Call script Draw function
        try
        {
            _jsEngine?.Invoke("Draw", gameTime.ElapsedGameTime.TotalSeconds);
        }
        catch (Exception ex)
        {
            MessageBox.Show($"Error in script Draw: {ex.Message}");
            Exit();
        }

        base.Draw(gameTime);
    }

    protected override void Dispose(bool disposing)
    {
        if (disposing)
        {
            _spriteBatch?.Dispose();
            _jsEngine?.Dispose();
        }

        base.Dispose(disposing);
    }

    private static void Main()
    {
        using GameHost host = new();
        host.Run();
    }
}